src/sentry/replays/usecases/reader.py

```diff
@@ -0,0 +1,256 @@
+            select=[Column("segment_id"), Column("retention_days"), Column("timestamp")],
+            where=[
+                Condition(Column("project_id"), Op.EQ, project_id),
+                Condition(Column("replay_id"), Op.EQ, str(uuid.UUID(replay_id))),
+                # We request the full 90 day range. This is effectively an unbounded timestamp
+                # range.
+                Condition(Column("timestamp"), Op.LT, datetime.now()),
+                Condition(Column("timestamp"), Op.GTE, datetime.now() - timedelta(days=90)),
+                # NOTE: Optimization to reduce the number of rows before the LIMIT clause. The
+                # cursors happen to map 1 to 1 with segment_id. If these filters are removed
+                # you will need to supply an offset value.
+                Condition(Column("segment_id"), Op.GTE, offset),
+                Condition(Column("segment_id"), Op.LT, offset + limit),
+                # Used to dynamically pass the "segment_id" condition for details requests.
+                *conditions,
+            ],
+            orderby=[OrderBy(Column("segment_id"), Direction.ASC)],
+            granularity=Granularity(3600),
+            limit=Limit(limit),
+            # NOTE: We do not use the offset parameter because of the segment_id query optimization
+            # in the where clause.  If you remove the segment_id filters you need to uncomment this
+            # offset value.
+            #
+            # offset=Offset(0),
+        ),
+    )
+    response = raw_snql_query(snuba_request, "replays.query.download_replay_segments")
+
+    return [
+        RecordingSegmentStorageMeta(
+            project_id=project_id,
+            replay_id=replay_id,
+            segment_id=item["segment_id"],
+            retention_days=item["retention_days"],
+            date_added=item["timestamp"],
+            file_id=None,
+        )
+        for item in response["data"]
+    ]
+
+
+# BLOB DOWNLOAD BEHAVIOR.
+
+
+def download_segments(segments: List[RecordingSegmentStorageMeta]) -> str:
+    """Download segment data from remote storage."""
+    with ThreadPoolExecutor(max_workers=4) as exe:
+        results = exe.map(download_segment, segments)
+        return (b"[" + b",".join(results) + b"]").decode("utf-8")
+
+
+def download_segment(segment: RecordingSegmentStorageMeta) -> bytes:
+    """Return the segment blob data."""
+    driver = FilestoreBlob() if segment.file_id else StorageBlob()
+    return decompress(driver.get(segment))
+
+
+def decompress(buffer: bytes) -> bytes:
+    """Return decompressed output."""
+    # If the file starts with a valid JSON character we assume its uncompressed. With time
+    # this condition will go extinct. Relay will compress payloads prior to forwarding to
+    # the next step.
+    if buffer.startswith(b"["):
+        return buffer
+
+    # Currently replays are gzipped by Relay.  Historical Replays retain their original gzip
```

Comment:
> Currently replays are gzipped by Relay. Historical Replays retain their original gzip compression from the SDK.
I'm not quite sure how to understand this comment -- the SDK is still certainly compressing replays, no? or do you just mean to say something like "replays regardless of status from the client client input are Gzipped by relay"?

Reply:
I agree.  Really unclear.  I'll update.